submissions/nadsatt/friends-app/modules/filter-items.js

```diff
@@ -0,0 +1,173 @@
+        this.name = `${property}-${category}`;
+        this.userService = userService;
+        this.defaultState = 'none';
+        this.defaultOptionElement = this.element.querySelector(`[data-state="${this.defaultState}"]`);
+        this.settedState = this.defaultState;
+        this.activatedOptionElement = this.defaultOptionElement;
+        this.classNames = {
+            openedFilterElement: 'filter-item--opened',
+            activatedOptionElement: 'filter-item__option--activated'
+        };
+    }
+
+    open(){
+        this.element.classList.add(this.classNames.openedFilterElement);
+    }
+
+    close(){
+        this.element.classList.remove(this.classNames.openedFilterElement);
+    }
+
+    setState(optionElement){
+        this.deactivateOptionAndIcon();
+        this.updateOptionAndState(optionElement);
+        if(this.settedState !== this.defaultState) this.activateOptionAndIcon();
+    }
+
+    applyAccordingToState(){
+        if(this.settedState !== this.defaultState){
+            this[`applyAccordingTo${this.getCapitalizedState(this.settedState)}State`]();
+        }
+    }
+
+    deactivateOptionAndIcon(){
+        this.activatedOptionElement.classList.remove(this.classNames.activatedOptionElement);
+        this.icon.deactivate();
+    }
+
+    activateOptionAndIcon(){
+        this.activatedOptionElement.classList.add(this.classNames.activatedOptionElement);
+        this.icon.activate();
+    }
+
+    updateOptionAndState(optionElement){
+        this.activatedOptionElement = optionElement;
+        this.settedState = optionElement.dataset.state;
+    }
+
+    getCapitalizedState(state){
+        return state[0].toUpperCase() + state.slice(1);
+    }
+}
+
+class SearchFilter extends Filter {
+    constructor(...args){
+        const optionsHTML =
+            `<input class="filter-item__option" type="text" data-state="search">`;
+        super(optionsHTML, ...args);
+
+        this.input = this.element.querySelector('input');
+    }
+
+    setState(optionElement){
+        this.deactivateOptionAndIcon();
+        this.updateOptionAndState(optionElement);
+
+        if(this.settedState === this.defaultState) this.input.value = ''
+        else if(this.input.value) this.activateOptionAndIcon();
+    }
+
+    applyAccordingToSearchState(){
+        const value = this.input.value.toLowerCase();
+
+        this.userService.users = this.userService.users.filter(user =>
+            user[this.property].toLowerCase().includes(value)
+        );
+    }
+}
+
+class ToggleFilter extends Filter {
+    constructor(firstState, secondState, ...args){
+        const optionsHTML =
+            `<span class="filter-item__option" data-state="${firstState}">${svgs[firstState]}</span>
+             <span class="filter-item__option" data-state="${secondState}">${svgs[secondState]}</span>`;
+        super(optionsHTML, ...args);
+
+        this.firstState = firstState;
+        this.secondState = secondState;
```

Comment:
Can this be less generic? Something like:

```
const states = {
 ascending: someValue,
 descending: ...
}
```

It is still unclear what two states you implying